# Comparative evaluation of bacterial pan-genome analysis tools
## Master's Thesis of Benjamin Otter

# results analysis
The formatted output files are located in the respective folders per species. The `.json` files contain the following information for each inferred cluster:
- geneId: cluster identification number 
- count: number of tags in this cluster
- dupli: yes/no flag for gene duplications
- sccg: yes/no flag for single copy core genes
- singleton: yes/no flag for singletons
- locus: list of locus tags in this cluster

### generate plots
All plots used in the Master's Thesis can be generated by running the `.ipynb` files.

### additional scripts
`cluster_distribution.py` maps the inferred clusters of one reference method to the clusters of all other methods for the same dataset. Every reference cluster is assigned to a distinct category relative to the method we compare to:
- split: tags of reference cluster are found in 2 or more clusters
- merge: reference cluster contains tags from 2 different clusters
- split: and merge: both of the above
- inflated: reference clusters contains tags not present in method we compare to
- incomplete: reference cluster is missing some tags present in the method we compare to
- other: Can't be assigned to any of the above categories
Parameters:
The dataset and organism names must match the folder structure and naming convention of the output files, raw data is exported as `.csv` file.
```
-d dataset
-o organism name
-r reference method
-p save plot
-e export raw data
```
Run the script:
```
python cluster_distribution.py -d Campy-180 -o Campy -r panXX -p -e
```


`cluster_to_tree.py` builds a tree based on protein sequences of a reference cluster and labels the tags with the results from another method. Sequences are aligned using MAFFT and the tree is built using Fasttree. 
Parameters:
```
-d dataset
-o organism name
-r reference method
-l labels method
-i geneId
```
Run the script:
```
python cluster_to_tree.py -d Campy-180 -o Campy -r panXX -l peppan -i 710
```
